libp2p/transports/tcptransport.nim

```diff
@@ -72,121 +72,135 @@ proc connHandler*(t: TcpTransport,
+      await client.closeWait()
 
-  return conn
-
-proc connCb(server: StreamServer,
-            client: StreamTransport) {.async, gcsafe.} =
-  trace "incoming connection", address = $client.remoteAddress
-  try:
-    let t = cast[TcpTransport](server.udata)
-    # we don't need result connection in this case
-    # as it's added inside connHandler
-    discard t.connHandler(client, false)
-  except CancelledError as exc:
     raise exc
-  except CatchableError as err:
-    debug "Connection setup failed", err = err.msg
-    client.close()
 
-proc init*(T: type TcpTransport, flags: set[ServerFlags] = {}): T =
+  return conn
+
+proc init*(T: type TcpTransport,
+           flags: set[ServerFlags] = {}): T =
   result = T(flags: flags)
+
   result.initTransport()
 
 method initTransport*(t: TcpTransport) =
   t.multicodec = multiCodec("tcp")
   inc getTcpTransportTracker().opened
 
-method close*(t: TcpTransport) {.async, gcsafe.} =
-  try:
-    ## start the transport
-    trace "stopping transport"
-    await procCall Transport(t).close() # call base
+method start*(t: TcpTransport, ma: MultiAddress) {.async.} =
+  ## listen on the transport
+  ##
 
-    checkFutures(await allFinished(
-      t.clients.mapIt(it.closeWait())))
+  if t.running:
+    trace "TCP transport already running"
+    return
 
-    # server can be nil
-    if not isNil(t.server):
-      t.server.stop()
-      await t.server.closeWait()
+  await procCall Transport(t).start(ma)
+  trace "Starting TCP transport"
 
-    t.server = nil
+  t.server = createStreamServer(t.ma, t.flags, t)
 
-    for fut in t.handlers:
-      if not fut.finished:
-        fut.cancel()
+  # always get the resolved address in case we're bound to 0.0.0.0:0
+  t.ma = MultiAddress.init(t.server.sock.getLocalAddress()).tryGet()
+  t.running = true
 
-    checkFutures(
-      await allFinished(t.handlers))
-    t.handlers = @[]
+  trace "Listening on", address = t.ma
 
-    for fut in t.cleanups:
-      if not fut.finished:
-        fut.cancel()
+method stop*(t: TcpTransport) {.async, gcsafe.} =
+  ## stop the transport
+  ##
+
+  try:
+    trace "Stopping TCP transport"
+    await procCall Transport(t).stop() # call base
 
     checkFutures(
-      await allFinished(t.cleanups))
-    t.cleanups = @[]
+      await allFinished(
+        t.clients[true].mapIt(it.closeWait()) &
+        t.clients[false].mapIt(it.closeWait())))
 
-    trace "transport stopped"
+    # server can be nil
+    if not isNil(t.server):
+      await t.server.closeWait()
+
+    t.server = nil
+    trace "Transport stopped"
     inc getTcpTransportTracker().closed
+  except CatchableError as exc:
+    trace "Error shutting down tcp transport", exc = exc.msg
+  finally:
+    t.running = false
+
+template withTransportErrors(body: untyped): untyped =
+  try:
+    body
+  except TransportTooManyError as exc:
+    warn "Could not create new connection, too many files opened", exc = exc.msg
+    raise newTransportClosedError(exc)
```

Comment:
`TransportTooManyError` is not fatal/critical error, most of the server software just note about this problem in logs, but it should not crash software. Because it is possible that on next accept free socket will be available.

Reply:
Noted, I'll make thin non fatal.